packages/discovery/src/discovery/output/saveDiscoveryResult.ts

```diff
@@ -63,6 +102,141 @@ export async function saveDiscoveryResult(
+  results: Analysis[],
+  logger: DiscoveryLogger,
+  options: {
+    rootFolder?: string
+    sourcesFolder?: string
+    discoveryFilename?: string
+  },
+): Promise<void> {
+  const flatSourcesFolder = `.flat${options.sourcesFolder ?? ''}`
+  const flatSourcesPath = posix.join(rootPath, flatSourcesFolder)
+  const allContractNames = results.map((c) =>
+    c.type !== 'EOA' ? c.name : 'EOA',
+  )
+
+  await rimraf(flatSourcesPath)
+
+  logger.log(`Saving flattened sources`)
+  for (const contract of results) {
+    try {
+      if (contract.type === 'EOA') {
+        continue
+      }
+
+      for (const [sourceIndex, source] of contract.source.entries()) {
+        const input: FileContent[] = Object.entries(source.files)
+          .map(([fileName, content]) => ({
+            path: fileName,
+            content,
+          }))
+          .filter((e) => e.path.endsWith('.sol'))
+
+        const result = timed(() => {
+          const parsedFileManager = ParsedFilesManager.parseFiles(
+            input,
+            source.remappings,
+          )
+          const output = flattenStartingFrom(source.name, parsedFileManager)
+
+          return output
+        })
+
+        const throughput = formatThroughput(input, result.executionTime)
+
+        const containingDirectory = getFlatContainingDirectoryName(
+          contract,
+          allContractNames,
+        )
+
+        const fileName = getFlatSourceFileName(
+          contract,
+          sourceIndex,
+          source,
+          allContractNames,
+        )
+
+        const path = posix.join(flatSourcesPath, containingDirectory, fileName)
+        await mkdirp(dirname(path))
+        await writeFile(path, result.value)
+
+        logger.log(`[ OK ]: ${source.name} @ ${throughput}/s`)
+      }
+    } catch (e) {
+      assert(contract.type !== 'EOA', 'This should never happen')
+      const contractName = contract.derivedName ?? contract.name
+
+      logger.log(`[FAIL]: ${contractName} - ${stringifyError(e)}`)
+    }
+  }
+}
+
+function getFlatContainingDirectoryName(
+  contract: Analysis,
+  allContractNames: string[],
+): string {
+  assert(contract.type !== 'EOA', 'Invalid execution path')
+
+  const hasNameClash =
+    allContractNames.filter((n) => n === contract.name).length > 1
+  const uniquenessSuffix = hasNameClash ? `-${contract.address.toString()}` : ''
+
+  const hasProxy = contract.source.length > 1
+  return hasProxy ? `${contract.name}${uniquenessSuffix}` : ''
+}
+
+function getFlatSourceFileName(
+  contract: Analysis,
+  sourceIndex: number,
+  source: PerContractSource,
+  allContractNames: string[],
+): string {
+  assert(contract.type !== 'EOA', 'Invalid execution path')
+
+  const hasProxy = contract.source.length > 1
+  const isProxy = hasProxy && sourceIndex === 0
+
+  const hasNameClash =
+    allContractNames.filter((n) => n === source.name).length > 1
+  const uniquenessSuffix =
+    hasNameClash && !hasProxy ? `-${source.address.toString()}` : ''
```

Comment:
It will never happen since proxies have a suffix `.p.sol`. So if there is a proxy named exactly the same as a contract the suffix will save it from clashing